### ISSUE TITLE
SC 1.3.5 Technique "Use type attribute on fields in simple login forms"?

### ISSUE THREAD

**Issue by detlevhfischer** [w3c/wcag#504]
In rating some forms against 1.3.5 Identify Input Purpose, I wondered whether the frequent two-field login forms that ask for just email and password would pass this SC already by specifying the type of the field (type=email, type=password). I reckon assistive technology presenting visual cues would not miss the opportunity to display the appropriate  image (not sure what these would be, though) to identify the purpose not just on forms that use the autocomplete attribute but also for (those parts of) forms that betray purpose in a programmatically determinable way via the type attribute.
So is there a case for a sufficient technique  "Use type attribute on fields in simple login forms", making it clear that it is scoped to just forms which contain no additional fields without unambiguous type?

**Comment by mraccess77** [comment]
We had this discussion on WebAIM recently.  The consensus from some seems to be that type can't be used.

https://webaim.org/discussion/mail_thread?thread=8936

**Comment by detlevhfischer** [comment]
Yes, I saw that discussion (and the epic exchange between JF and Jared) but I still feel that in a narrowly constrained context such as login, where it is clear that just you email and your password are needed, type should suffice to meet the SC. Why should cogs AT not be smart enough to look out for type as well, in the absence of autocomplete and supply fitting icons when there is an unambiguous match? Once you also have other fields like name that would be of type text, this obviously stops working. But I am talking here about a very common, restricted and conventional scenario.

**Comment by patrickhlauke** [comment]
i would say `type="email"` and `type="password"`, in isolation and without any other fields of the same type, should likely be treated as acceptable for a login (compared to, say, two fields just with `type="text"`. i'd say it's pretty unambiguous what purpose those two inputs would have.

though i could see a more pedantic reading that would say the only acceptable way here would be:
```
<input type="email" ...>
<input type="password" autocomplete="current-password">
```

**Comment by detlevhfischer** [comment]
@patrickhlauke ....which raises the interesting question what you’d have to do in the frequent case that an input field (just like the one I just used to login to Github) accepts BOTH user name AND email. Is it valid for autocomplete to have two values? I guess not (but too lazy to check). So you would, when using autocomplete, misrepresent the purpose either way, simply by omission. What would be the right thing to do on those form fields, and in turn, the right approach in conformance testing?

**Comment by patrickhlauke** [comment]
heh, i was thinking about that last night when answering, but didn't want to open that particular can of worms. my reading of https://www.w3.org/TR/html52/sec-forms.html#sec-autofill seems to suggest that you can't assign multiple `autocomplete` values (and i'm sure somewhere in the complex algorithm described in HTML5.2 there's something relating to how to handle tokens that refer to different purposes being present for the same control, but the techno-babble of that part confuses me). i think the short answer would be that the input has no set/defined purpose, as it's essentially a multi-purpose freeform text entry (which the server then compares to usernames and email addresses), which with current technologies can't be defined unambiguously and is therefore currently exempt from pass/fail.

**Comment by patrickhlauke** [comment]
(which then raises the question: what happens when technologies do become available to allow things? you'd have to retrospectively pass/fail things that before were N/A. this feels like an "until user agents..." kind of situation in reverse...)

**Comment by alastc** [comment]
Discussed and closed... Detlev still keen to see examples where there are multiple people being added to something.

**Comment by patrickhlauke** [comment]
Sorry to dig this up again, but: what was the agreed consensus? That it's kosher to just rely on `type="email"` / `type="password"` in forms (where UA/AT would likely infer this refers to the person's email/password?

If so, this part of the understanding doc https://www.w3.org/WAI/WCAG21/Understanding/identify-input-purpose.html will likely need to be changed accordingly:

> For some input fields, the type attribute already offers a way to broadly specify the intention of the input field, for example, input type="tel", input type="email", or input type="password". However, these are only very broad categories, describing the type of input, but not necessarily its purpose, especially as it relates to user-specific input fields. As an example, type="email" indicates that the field is for an e-mail address but does not clarify if the purpose is for entering the user's e-mail address or some other person's e-mail.

Related, if the answer/decision above was that `type="email"` and `type="password"` are acceptable, what about `type="tel"` and `type="url"` ? in a simple form, can they be assumed to be the phone/web address for the user themselves?

**Comment by detlevhfischer** [comment]
@patrickhlauke I think the discussion at the time concluded that even in such a restricted case (login with just email and password), `type` would not do the trick as new AT would more likely latch onto some unique purpose identifier (for the time being, `autocomplete`). I remain somewhat unconvinced but accepted that resolution. The text you quote draws on a distinction between 'intention' and 'purpose' which I find somewhat unconvincing - maybe one should better argue that `type` sets technical constraints for inputs and cannot be extended to cover all input purposes defined in [part 7 of WCAG 2.1](https://www.w3.org/TR/WCAG21/#input-purposes), so for generality's sake (?) it should not be used even in restricted cases where it might suffice.

**Comment by jake-abma** [comment]
that's also how I recall it, only type wouldn't do the trick and requires some heuristic analysis to be sure this is for example a login screen yes/no. Conclusion was: "not kosher"

**Comment by patrickhlauke** [comment]
fair enough. have to admit that this kind of speculative requirement (as there's currently nothing, to my knowledge, that really takes full advantage of autocomplete in this way) still rubs me the wrong way, as it's pretty much always a fail in audits that i do unless the authors explicitly added these (to pass the SC for the most part)...but at least it's clear.

**Comment by alastc** [comment]
The issue was essentially false positives, i.e. it is not unusual to have forms with multiple names, but only one of which is aimed at being *your* name. See #642 for even more detail than you probably want.

**Comment by patrickhlauke** [comment]
dropping this here, but worth noting: apparently i was wrong, and a form field CAN have multiple autocomplete tokens (while i knew the spec said it allows an ordered list of tokens, i was under the impression this referred more to the fact that there are optional tokens that can be there as well)

https://twitter.com/patrick_h_lauke/status/1130468323060342784

long story short, for that case of inputs that allow either username or email, it seems that it's valid (though possibly not supported) to use `autocomplete="username email"` or similar.

**Comment by patrickhlauke** [comment]
ah, wonderful ... there's disagreement between spec writers on this point. good times... https://twitter.com/annevk/status/1130478430313746432

**Comment by patrickhlauke** [comment]
ignore the above. seems the spec is just vague, but that our original reading (that it does NOT allow these sorts of multiple autocomplete values) was correct after all. as you were...

**Comment by detlevhfischer** [comment]
@patrickhlauke @johnfoliot Sorry to pick this one up once again: Since we have established that two separate values  (like email and username) are not legit (except for purposes of bounding values to contexts like shipping and billing)  I assume that on a typical login field that allows input of **either** username **or** email, there would be two options and BOTH would pass:
- not using autocomplete here at all because the field is not unambiguous
- using either username OR email as value of autocomplete

Agree?

**Comment by patrickhlauke** [comment]
i would, yes

**Comment by mraccess77** [comment]
What is the final outcome of whether autocomplete is required on fields that take either username or email?

**Comment by patrickhlauke** [comment]
flip a coin and pick one?

**Comment by detlevhfischer** [comment]
Proposed working with answer:
In cases where a field accepts  two types of entry / two values for user purpose (as in combined name/password fields), it is valid to give either one or the other value or leave out the designation of input purpose altogether.

**Comment by alastc** [comment]
The working agreed with this response:

> In cases where a field accepts two types of entry / two values for user purpose (as in combined user name/user email fields), it is valid to give either one or the other value or leave out the designation of input purpose altogether. We have updated the text of 1.3.4 "Understanding Input Purpose" to cover this situation.

**Comment by alastc** [comment]
@detlevhfischer is the understanding update you refered to somewhere I can see?

**Comment by detlevhfischer** [comment]
@alastc yes, it is here https://github.com/w3c/wcag/pull/1217
